### PR TITLE
Fixes multiple reliability-Problems around Pad.block()

### LINKED ISSUE #184: Reliability-Problems around Pad.block()
I think I might be onto a race condition in `Pad.block()`. I'm not yet able to reproduce the problem outside of my application, but i can reproduce it about 50% of the time there.

The relevant Code-Section looks like this:
```java
Pad teePad = …;
CompletableFuture<Void> future = new CompletableFuture<>();
log.info("requesting block of teePad {} of {} - isBlocked={} isBlocking={}",
		teePad.getName(), tee.getName(), teePad.isBlocked(), teePad.isBlocking());

GstDebugger.debugPipeline("before-block", pipeline);
teePad.block(() -> {
	log.info("blocked teePad {} of {} - isBlocked={} isBlocking={}",
			teePad.getName(), tee.getName(), teePad.isBlocked(), teePad.isBlocking());

	// release pad here

	future.complete(null);
});

log.info("after requesting block of teePad {} of {} - isBlocked={} isBlocking={}",
		teePad.getName(), tee.getName(), teePad.isBlocked(), teePad.isBlocking());

GstDebugger.debugPipeline("after-block", pipeline);
future.get();
```
A src-Pad of a tee is blocked in order to correctly release it.
If the pad is currently idle, one would expect the Runnable-Closure to be called immediately, otherwise one would expect the call to `.block` to complete and the callback being called shortly thereafter.

50% of the time this work and the logs look like this:
```
[main] INFO de.mazdermind.gintercom.mixingcore.Panel - requesting block of teePad src_0 of rx-1 - isBlocked=false isBlocking=false
[main] INFO de.mazdermind.gintercom.mixingcore.Panel - after requesting block of teePad src_0 of rx-1 - isBlocked=true isBlocking=false

(gst1-java-core:14957): GStreamer-CRITICAL **: 18:27:10.478: gst_pad_probe_info_get_event: assertion 'info->type & (GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM | GST_PAD_PROBE_TYPE_EVENT_UPSTREAM)' failed
[GstBus] INFO de.mazdermind.gintercom.mixingcore.Panel - blocked teePad src_0 of rx-1 - isBlocked=true isBlocking=false

(gst1-java-core:14957): GStreamer-CRITICAL **: 18:27:10.480: gst_pad_probe_info_get_event: assertion 'info->type & (GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM | GST_PAD_PROBE_TYPE_EVENT_UPSTREAM)' failed
```
aside from the assertation, which is a different issue and probably not harmful, one can see that the callback has been called from the GstBus thread and that right after the request to block, the pad is already blocked.

The other 50% of the time
the logs show, that the pad is still being blocked, but the callback never gets called:
```
[main] INFO de.mazdermind.gintercom.mixingcore.Panel - requesting block of teePad src_0 of rx-1 - isBlocked=false isBlocking=false
[main] INFO de.mazdermind.gintercom.mixingcore.Panel - after requesting block of teePad src_0 of rx-1 - isBlocked=true isBlocking=false
…
(Timeout in Test-Runner after some Minutes)
```

The dot-File-Diagrams show now surprises; the Pad is displayed as blocked just as the result of `isBlocked` suggest: https://mazdermind.de/volatile/dot-files.zip

*Edit* I can reproduce the issue with gst1-java-core 1.1.0 and 1.2.0-SNAPSHOT from master

### PR DESCRIPTION
**Pull request by MaZderMind**
This PR fixes multiple Reliability-Problems around Pad.block() that are descussed in great length in #184. The PR contains multiple fixes and a new Api, `Pad.blockAndWait`, so If you want I can split it up into multiple PRs. The different fixes and changes are in different well separated commits though, so maybe that's enough.

fixes #184

### CODE REVIEW

**Comment by MaZderMind** [comment]
rebased to master

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+# Unix-style newlines with a newline ending every file
+[*]
+end_of_line = lf
+insert_final_newline = true
+indent_style = space
+indent_size = 4
+charset = utf-8
+
+[*.xml]
+indent_size = 2
```
Comment (neilcsmith-net) [inline .editorconfig:13]: Not against this.  Useful for IntelliJ? Possibly move to separate PR along with `.gitignore`

File: .gitignore [hunk 2]
```diff
@@ -16,3 +16,4 @@
 nb-configuration.xml
 
 .idea
+*.iml
```
Comment (neilcsmith-net) [inline .gitignore:19]: As above

File: src/org/freedesktop/gstreamer/Pad.java [hunk 3]
```diff
@@ -318,9 +328,45 @@ public void block(final Runnable callback) {
             public PadProbeReturn eventReceived(Pad pad, Event event) {
                 callback.run();
                 pad.removeCallback(EVENT_PROBE.class, this);
-                return PadProbeReturn.DROP;
+                return PadProbeReturn.REMOVE;
             }
-        }, GstPadAPI.GST_PAD_PROBE_TYPE_BLOCKING | GstPadAPI.GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM);
+        }, GstPadAPI.GST_PAD_PROBE_TYPE_IDLE);
```
Comment (neilcsmith-net) [inline src/org/freedesktop/gstreamer/Pad.java:333]: I'm concerned this is changing the semantics of an existing method. Even if it currently has some issues. I'm not against it, as it's probably the right change. Are there any uses that might break with this change?

I am inclined to fix `block()` but also deprecate it with a view to adding generic `addProbe()` support. The idea of gst1-java-core is to stick close to implementing upstream GStreamer behaviour rather than add to it.

Having said that, wonder whether to deprecate `block()` and add a new method with your fixes - `onIdle(Runnable r)` ?

File: src/org/freedesktop/gstreamer/Pad.java [hunk 4]
```diff
@@ -318,9 +328,45 @@ public void block(final Runnable callback) {
...
+     * The pad will remain blocked for as long as the callback is active.
+     * The callback may be called from this thread or another thread.
+     *
+     * @param callback The code to run when pad is blocked
+     */
+    public void blockAndWait(final Runnable callback) {
+        blockAndWait(() -> {
+            callback.run();
+            return null;
+        });
```
Comment (neilcsmith-net) [inline src/org/freedesktop/gstreamer/Pad.java:369]: I'd rather not add these. They are things better implemented outside of the core library, and not sure about the hardcoded assumptions - eg. timeout. Probably better achieved by passing a `FutureTask` into `block()` / `onIdle()` and calling `get()` explicitly.

File: src/org/freedesktop/gstreamer/Pad.java [hunk 5]
```diff
@@ -390,12 +436,14 @@ public void addEventProbe(final EVENT_PROBE listener) {
...
+    synchronized void addEventProbe(final EVENT_PROBE listener, final int mask) {
         final GstPadAPI.PadProbeCallback probe = new GstPadAPI.PadProbeCallback() {
             public PadProbeReturn callback(Pad pad, GstPadProbeInfo probeInfo, Pointer user_data) {
-//        	    System.out.println("CALLBACK " + probeInfo.padProbeType);
                 if ((probeInfo.padProbeType & mask) != 0) {
-                    Event event = GSTPAD_API.gst_pad_probe_info_get_event(probeInfo);
+                    Event event = null;
+                    if ((probeInfo.padProbeType & EVENT_HAS_INFO_MASK) != 0) {
+                        event = GSTPAD_API.gst_pad_probe_info_get_event(probeInfo);
+                    }
```
Comment (neilcsmith-net) [inline src/org/freedesktop/gstreamer/Pad.java:446]: Looks good!

File: src/org/freedesktop/gstreamer/Pad.java [hunk 6]
```diff
@@ -431,7 +486,7 @@ public PadProbeReturn callback(Pad pad, GstPadProbeInfo probeInfo, Pointer user_
             }
         };
 
-        GCallback cb = new GCallback(handle.addProbe(GstPadAPI.GST_PAD_PROBE_TYPE_BUFFER, probe), probe) {
+        GCallback cb = new GCallback(handle.addProbe(GstPadAPI.GST_PAD_PROBE_TYPE_BUFFER, probe, this), probe) {
```
Comment (neilcsmith-net) [inline src/org/freedesktop/gstreamer/Pad.java:489]: We can't do this - see later comment!

File: src/org/freedesktop/gstreamer/Pad.java [hunk 7]
```diff
@@ -660,34 +715,42 @@ public boolean hasCurrentCaps() {
...
         }
 
         @Override
         protected GstPadPtr getPointer() {
             return (GstPadPtr) super.getPointer();
         }
-        
-        private synchronized NativeLong addProbe(int mask, GstPadAPI.PadProbeCallback probe) {
+
+        private synchronized NativeLong addProbe(int mask, GstPadAPI.PadProbeCallback probe, Pad pad) {
```
Comment (neilcsmith-net) [inline src/org/freedesktop/gstreamer/Pad.java:734]: Sorry, but we can't keep reference to `Pad` in `Pad.Handle`. This will cause memory leaks, even if it fixes your immediate problem.

File: src/org/freedesktop/gstreamer/glib/GObject.java [hunk 8]
```diff
@@ -588,7 +588,7 @@ private static void transform(Object data, GType type, GValue dst) {
         protected GCallback(NativeLong id, Callback cb) {
             this.id = id != null ? id : new NativeLong(0);
             this.cb = cb;
-            this.connected = this.id.intValue() != 0;
+            this.connected = this.id.longValue() != 0;
```
Comment (neilcsmith-net) [inline src/org/freedesktop/gstreamer/glib/GObject.java:591]: Looks good. Was this actually causing any issue?

File: .editorconfig (hunk 1 above)
Comment (MaZderMind) [reply .editorconfig:13]: This is used by mose IDEs nowadays, IntelliJ, VSCode and Eclipse (probably with a Plugin( at least. Which IDE are you using?

File: src/org/freedesktop/gstreamer/glib/GObject.java (hunk 8 above)
Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/glib/GObject.java:591]: No, just Optimization along the way

File: src/org/freedesktop/gstreamer/Pad.java (hunk 4 above)
Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/Pad.java:369]: I think that this is a fairly common use-case, for example when changing the Pipeline from a GUI Thread or from a Network Thread. In C there is no well defined way to synchronize these Paths, but in Java there is.

The hardcoded Timeout should probably better be a Parameter with a Default-Overload.

File: src/org/freedesktop/gstreamer/Pad.java (hunk 7 above)
Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/Pad.java:734]: I had a Test-Program (see Gist in the original Issue) spin around .block() a couple million times and monitored the stack with VisualVM, and there is no memory-leak here.

The String Ref to the Pad is stored alongside with the Probe-Callback and removed when the Probe-Callback is removed. The GC is then free to clean the Pad up. The reference in Handle is only held as long as there is a Probe pending.

**Comment by MaZderMind** [comment]
I'm fine with splitting the PR into multiple. Let's finish the discussion here and then I'll close this one and open individual PRs per group of changes.

File: src/org/freedesktop/gstreamer/Pad.java (hunk 7 above)
Comment (neilcsmith-net) [reply src/org/freedesktop/gstreamer/Pad.java:734]: I'm concerned about other methods than `block()`.

File: src/org/freedesktop/gstreamer/Pad.java (hunk 3 above)
Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/Pad.java:333]: According to the Javadoc, `.block` does "Run a runnable under a blocked state".
GST_PAD_PROBE_TYPE_IDLE does that as well as GST_PAD_PROBE_TYPE_BLOCKING with the only difference, that GST_PAD_PROBE_TYPE_BLOCKING will only be called as soon as there is actual data flowing on a Pad (and the Data-Pushing Thread is then calling the Probe) whereas GST_PAD_PROBE_TYPE_IDLE can also be called directly from the probe-add.

The implication of this is, that a Probe for GST_PAD_PROBE_TYPE_BLOCKING might *never* be called, when there is no actual dataflow (ie. the source is some kind of network source, lets say an udpsrc) and you want to unlink it, then blocking for GST_PAD_PROBE_TYPE_BLOCKING might not work if the sending side does not send anymore packets and your probe will never be called.

Before GST_PAD_PROBE_TYPE_IDLE was introduced, one had to manually inject an event into the pipeline force a callback for GST_PAD_PROBE_TYPE_BLOCKING even when no data is flowing.

To to answer your question: I think that only blocking for GST_PAD_PROBE_TYPE_IDLE does exactly what the Javadoc and Method-Name indicates and what the caller expects.

File: src/org/freedesktop/gstreamer/Pad.java (hunk 4 above)
Comment (neilcsmith-net) [reply src/org/freedesktop/gstreamer/Pad.java:369]: It's a fairly common way to cause problems too! :smile: And encouraging people to block their GUI or network thread is not high on my agenda.  Considering people can easily do, if they really need this and know there's no deadlock -

```
var task = new FutureTask(() -> connectSomething() ), null);
pad.block(task);
task.get(40, TimeUnit.DAYS);
```

File: src/org/freedesktop/gstreamer/Pad.java (hunk 3 above)
Comment (neilcsmith-net) [reply src/org/freedesktop/gstreamer/Pad.java:333]: I think you're mixing up `GST_PAD_PROBE_TYPE_BLOCKING` and `GST_PAD_PROBE_TYPE_BLOCK`. Now, I'm not sure if anyone could be relying on the behaviour of the former.

File: src/org/freedesktop/gstreamer/Pad.java (hunk 7 above)
Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/Pad.java:734]: As the [Probe-Callback](https://github.com/gstreamer-java/gst1-java-core/blob/master/src/org/freedesktop/gstreamer/Pad.java#L395) (not only for block but for all Probes) refers to the Pad, the Pad-Instance *has* to live as long as the Probe does. When the Probe goes away, it *has* to call `Handle.removeProbe` (it does so through `GCallback.disconnect` when the Callback is removed).

Thus I cannot see, how the Pad-Instance in the Handle could form a Memleak without the Probe also forming one. It should be fairly simple to generate a small piece of Code that exposes such a leak ant it should be fairly easy to spot in VisualVM.

Comment (neilcsmith-net) [reply src/org/freedesktop/gstreamer/Pad.java:734]: You mean because it's an inner class (not the method argument)?  That the Pad and Probe reference each other isn't a problem. But `Handle` should not under any circumstances have a hard reference to the object it's a handle for.  `Handle` has a very particular job to do around memory handling, and this risks bringing in huge problems.

File: src/org/freedesktop/gstreamer/Pad.java (hunk 3 above)
Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/Pad.java:333]: You are correct, I mixed them up in my comment.

In the usage `GstPadAPI.GST_PAD_PROBE_TYPE_BLOCKING | GstPadAPI.GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM` the Mask did match both, BLOCK and IDLE, which would then possibly call the Runnable twice.

File: src/org/freedesktop/gstreamer/Pad.java (hunk 4 above)
Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/Pad.java:369]: Ok, I can see that this API might lead to unexpected Uses. I'm fine with removing it.
I just found that I wrote the same Boilerplate in multiple places around `.block` and chose to add it; but I can see why you do not want this to be part of the API Bindings. I'll remove it from the splitted PRs.

File: src/org/freedesktop/gstreamer/Pad.java (hunk 3 above)
Comment (neilcsmith-net) [reply src/org/freedesktop/gstreamer/Pad.java:333]: Yes, my thought too, and why is the event mask there?!  This was added before I took on the fork, but it does seem wrong - let's go with your change here and if anyone complains we can rethink it!

File: src/org/freedesktop/gstreamer/Pad.java (hunk 7 above)
Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/Pad.java:734]: I see that I mis-used (and did not understand) Handle here. Let's talk about alternatives. I though about implementing the Probe-Object as a Closure that holds an implicit reference to the surrounding Pad; something like
```
Pad padRef = this;
final GstPadAPI.PadProbeCallback probe = new GstPadAPI.PadProbeCallback() {
   …
   return listener.eventReceived(padRef, event);
   …
}
```
This should also keep a Reference, as long as the inner Probe-Object exists. I have to check this, though.

Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/Pad.java:734]: I tried that but it alone does not help. To copy from the Issue (and remind myself of the actual problem): The root cause is, that
 - the GstPadAPI.PadProbeCallback instance is only held on the Pad object
 - nothing guarantees that the Pad object lives long enough so that not Path, the Pad and the GstPadAPI.PadProbeCallback instance are GCed while the Probe is in place.

So somewhere the Pad *OR* the GstPadAPI.PadProbeCallback instance needs to be stored *within the Java Object-Graph* while the Probe is active *within the GStreamer Stack*.

*OR*, because we can attach a Reference to the Pad to the GstPadAPI.PadProbeCallback by making it a final reference as mentioned above.

Comment (neilcsmith-net) [reply src/org/freedesktop/gstreamer/Pad.java:734]: Handle is part of a dual hierarchy that is used in NativeObject.NativeRef - https://github.com/gstreamer-java/gst1-java-core/blob/master/src/org/freedesktop/gstreamer/glib/NativeObject.java#L241  Note that that is a weak reference to the native object with a strong reference to the handle - it's how the memory handling works and the handle having a strong reference back to the actual native object is likely to cause problems.

My inclination may still be to store the Pads inside the Element where necessary.  But let's clarify where exactly the Pad is coming from, and the ownership / refcounts on it?  I have some code I found that tests some probe calls for GC issues from earlier problems - not sure when - but that doesn't have an issue.

Comment (MaZderMind) [reply src/org/freedesktop/gstreamer/Pad.java:734]: The actual Code triggering the Problem reliably is this: https://github.com/MaZderMind/gintercom-mixing-core/blob/master/src/main/java/de/mazdermind/gintercom/mixingcore/Panel.java#L123-L132

As yo ucan see, the Pad-Object is generated from `Pad mixerPad = ((GhostPad) pad).getTarget();`

Comment (neilcsmith-net) [reply src/org/freedesktop/gstreamer/Pad.java:734]: Can you see if adding `@CallerOwnsReturn` to `gst_element_get_request_pad` helps?   eg. https://github.com/gstreamer-java/gst1-java-core/blob/master/src/org/freedesktop/gstreamer/lowlevel/GstElementAPI.java#L77 doesn't match up with `transfer-full` at https://developer.gnome.org/gstreamer/stable/GstElement.html#gst-element-get-request-pad

**Comment by MaZderMind** [comment]
I extracted all changes related to the reliability of `Pad.block` into their respective PRs; closing this PR.
